### ISSUE TITLE
Repeating Multiple Tasks

### ISSUE THREAD

**Issue by nickm2402** [PrefectHQ/prefect#975]
Hello,

I have been playing around with prefect and am really happy with it, but currently stuck with an issue. 

I have the following flow of Tasks A to E:
A->B->C->D->E

If D fails, I need to repeat B->C->D. How can I achieve that with Prefect? What is best practice in that scenario?

Do I need to split it into three flows? 1(A), 2(B->C->D), 3(E) and repeat the second flow?

Do I need to make a bigger Task (BCE) that combines smaller tasks?

I have read the discussion in #674, but since it was declined I am not sure what the best way to solve these kinds of issues is.

Thanks for any hint in the right direction.

**Comment by jlowin** [comment]
Hey Nick — this is a really interesting question!

Prefect has a strong assumption that once a task finishes (in a given run of the flow), it won’t run again. Let’s say you had a task F that followed C. And let’s say that F ran successfully after C succeeded. If D caused C to run a second time, then the behavior of F is sort of undefined - should it necessarily retry as well? If not, what if its state is now inconsistent with C? If yes, what does this mean about Prefect’s ability to guarantee run-once semantics (because tasks are causing other tasks to run again)

To be clear, I’m not trying to lay out this this is *bad* - just showing why solving it in a general way introduces a ton of complexity, which is why you’re encountering a little friction in expressing it with Prefect. Generally, the system don’t like it when tasks affect any other tasks but their immediate downstream neighbors. 

Would B-C-D naturally represent one unit of work in your flow? If so, I would encourage combining them. It’ll solve your retry problem and comply with your workflow semantics. 

Your three flows idea would work as well, but create some overhead.

Lastly, perhaps this motivates a slightly different approach than the declined PIN you mentioned. Pinging @cicdw for any thoughts.

**Comment by cicdw** [comment]
Great question; ultimately, the Flow as you have presented is not a DAG.  However, there are various ways to achieve the goal:
- combine B-C-D into a single Task.  This is probably the simplest approach to achieve what you want
- Break your workflow into 3 individual flows.  This has a lot of overhead and might not be satisfying for your use case (it sounds like your process is naturally thought of as a single workflow).  Additionally, this introduces the possibility that you forget which runs of A are associated with which runs of B->C->D which is undesirable.
- there is a third approach that comes to my mind, but I'm not sure if it's the right approach -> you could _cache_ task A based on time or inputs, and then if D fails you could simply manually re-run the whole flow.  This would have the effect of not actually rerunning A, then rerunning B->C->D.  The only downside is constructing the appropriate caching logic for A + the fact that you'd most likely have to _manually_ rerun the whole Flow

**Comment by nickm2402** [comment]
Thank you for your replies.

I don't really want to combine B, C and D if I can avoid it. They can work separately and should stay as units if possible.

I oversimplified my case to explain my issue, so breaking it up into more flows might overcomplicate things. I am going to try the third approach and use caching. This is something I wanted to experiment with anyway.

Only slightly on topic, but I am curious. You allow for retries of single tasks. Retrying multiple tasks by throwing everything away up to a specific point and retrying from there again seems to be the same in terms of the graph being cyclic/acyclic. Doesn't even a single task being rerun/retried makes it cyclic too (strictly speaking) and breaks the philosophy of running tasks only once?

**Comment by cicdw** [comment]
It's a good question; there isn't really a loop there because the task is never considered "finished" until it is done retrying.  In the multiple task retry situation, the tasks would be considered _finished_, and then rerun, and that's where the contract gets violated.

Glad to hear caching might be a good solution for you - let us know how it goes!

**Comment by cicdw** [comment]
Closing as the use case has potentially been resolved via caching.
